contracts/mainnet/connectors/uniswapV3/main.sol

```diff
@@ -0,0 +1,184 @@
+        amountB = getUint(getIds[1], amountB);
+
+        (uint256 _liquidity, uint256 _amtA, uint256 _amtB) = _addLiquidity(
+            tokenId,
+            amountA,
+            amountB,
+            slippage
+        );
+        setUint(setId, _liquidity);
+
+        _eventName = "LogDeposit(uint256,uint256,uint256,uint256)";
+        _eventParam = abi.encode(tokenId, _liquidity, _amtA, _amtB);
+    }
+
+    /**
+     * @dev Decrease Liquidity
+     * @notice Decrease Liquidity of NFT Position
+     * @param tokenId NFT LP Token ID.
+     * @param liquidity LP Token amount.
+     * @param amountAMin Min amount of tokenA.
+     * @param amountBMin Min amount of tokenB.
+     * @param getId ID to retrieve LP token amounts
+     * @param setIds stores the amount of output tokens
+     */
+    function withdraw(
+        uint256 tokenId,
+        uint256 liquidity,
+        uint256 amountAMin,
+        uint256 amountBMin,
+        uint256 getId,
+        uint256[] calldata setIds
+    )
+        external
+        payable
+        returns (string memory _eventName, bytes memory _eventParam)
+    {
+        if (tokenId == 0) tokenId = _getLastNftId(address(this));
+        uint128 _liquidity = uint128(getUint(getId, liquidity));
+
+        (uint256 _amtA, uint256 _amtB) = _decreaseLiquidity(
+            tokenId,
+            _liquidity,
+            amountAMin,
+            amountBMin
+        );
+
+        setUint(setIds[0], _amtA);
+        setUint(setIds[1], _amtB);
+
+        _eventName = "LogWithdraw(uint256,uint256,uint256,uint256)";
+        _eventParam = abi.encode(tokenId, _liquidity, _amtA, _amtB);
+    }
+
+    /**
+     * @dev Collect function
+     * @notice Collect from NFT Position
+     * @param tokenId NFT LP Token ID.
+     * @param amount0Max Max amount of token0.
+     * @param amount1Max Max amount of token1.
+     * @param getIds IDs to retrieve amounts
+     * @param setIds stores the amount of output tokens
+     */
+    function collect(
+        uint256 tokenId,
+        uint256 amount0Max,
+        uint256 amount1Max,
+        uint256[] calldata getIds,
+        uint256[] calldata setIds
+    )
+        external
+        payable
+        returns (string memory _eventName, bytes memory _eventParam)
+    {
+        if (tokenId == 0) tokenId = _getLastNftId(address(this));
+        uint128 _amount0Max = uint128(getUint(getIds[0], amount0Max));
+        uint128 _amount1Max = uint128(getUint(getIds[1], amount1Max));
+        (uint256 amount0, uint256 amount1) = _collect(
+            tokenId,
+            _amount0Max,
+            _amount1Max
+        );
+
+        setUint(setIds[0], amount0);
+        setUint(setIds[1], amount1);
+        _eventName = "LogCollect(uint256,uint256,uint256)";
+        _eventParam = abi.encode(tokenId, amount0, amount1);
+    }
+
+    /**
+     * @dev Burn Function
+     * @notice Burn NFT LP Position
+     * @param tokenId NFT LP Token ID
+     */
+    function burn(uint256 tokenId)
```

Comment:
burn will only happen if the NFT is empty. Right?

Reply:
Sure